src/psyclone/psyir/backend/sympy_writer.py

```diff
@@ -100,49 +103,122 @@ def __init__(self, list_of_expressions=None):
+        '''
+        This function creates a dictionary mapping each Reference in any
+        of the expressions to either a Sympy Function (if the reference
+        is an array reference) or a Symbol (if the reference is not an
+        array reference).
+
+        :param list_of_expressions: the list of expressions from which all \
+            references are taken and added to the a symbol table to avoid \
+            renaming any symbols (so that only member names will be renamed).
+        :type list_of_expressions: list of \
+            :py:class:`psyclone.psyir.nodes.Node`
+        :returns: the dictionary mapping each reference name to a Sympy \
+            data type (Function of Symbol).
+        :rtype: dictionary of string:Sympy-data-type values
+
+        '''
+        sympy_type_map = {}
+        for expr in list_of_expressions:
+            for ref in expr.walk(Reference):
+                name = ref.name
+                if name not in sympy_type_map:
+                    if ref.is_array:
+                        sympy_type_map[name] = Function(name)
+                    else:
+                        sympy_type_map[name] = Symbol(name)
+
+        return sympy_type_map
 
+    @staticmethod
+    def convert_to_sympy_expressions(list_of_expressions):
         '''
-        return self._sympy_type
+        This function takes a list of PSyIR expressions, and converts
+        them all into Sympy expressions using the SymPy parser.
+        It takes care of all Fortran specific conversion required (e.g.
+        constants with kind specification, ...), including the renaming of
+        member accesses, as described in
+        https://psyclone-dev.readthedocs.io/en/latest/sympy.html#internal-details
+
+        It returns the converted PSyIR expressions as a list of SymPy
+        expressions.
+
+        :param list_of_expressions: the list of expressions which are to be \
+            converted into SymPy-parsable strings.
+        :type list_of_expressions: list of \
+            :py:class:`psyclone.psyir.nodes.Node`
+        :returns: a list of SymPy expressions that are the conversion from
+            the specified PSyIR expressions.
+        :rtype: list of SymPy expressions
+
+        '''
+        # Create the type_map that will include all symbols used in both
+        # expressions.
+        type_map = SymPyWriter.create_type_map(list_of_expressions)
+
+        # Create a SymPy writer that uses this type_map
+        writer = SymPyWriter(type_map)
+        expression_str_list = []
+        for expr in list_of_expressions:
+            # Convert each expression. Note that this call might add
+            # additional entries to type_map if it finds member names
+            # that clash with a symbol (e.g. a%b --> it will try to
+            # create a SymPy symbol `a_b`, but if `a_b` is the same as
+            # symbol, `a_b_1`, ... will be used instead).
+            expression_str_list.append(writer(expr))
```

Comment:
I just tried, but at this stage sympy does not handle array expressions (basically anything with a : in it - #1537). I tried some rough replacing of : expressions with constants, but ended up with various Fortran problems and/or psyclone crashes (invalid Fortran ... using where or elsewhere without an actual array expression:
```
../psyir/frontend/fparser2.py:3852: in _module_handler
    self.process_declarations(container, spec_part.children,
../psyir/frontend/fparser2.py:1975: in process_declarations
    self._process_decln(parent, parent.symbol_table, node,
>       return root_loop
E       UnboundLocalError: local variable 'root_loop' referenced before assignment

../psyir/frontend/fparser2.py:3070: UnboundLocalError
```
Well, I did have some limited success: it takes a bit over 3 seconds to parse this file without doing anything. I then added:
```
    for i in psyir.walk(ArrayReference):
        print("I", i)
        assert sym_maths.equal(i, i) is True
```
So basically comparing any ArrayReference with itself, and it took 31 seconds to process 58 expressions before it crashed (due to an array expression). When I used `writer._visit(...)` instead of just `writer(...)` ... the time went down to 3.7 seconds. Gee, HUGE difference.

Since you have used the _visit work around, I will use the same here then! Very good catch, thanks !!

Reply:
Great. Please could you add a TODO for #1587 here too.

Reply:
Done.